staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go

```diff
@@ -0,0 +1,186 @@
+	flowcontrolClient fcclientv1a1.FlowcontrolV1alpha1Interface,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	queueSetFactory fq.QueueSetFactory,
+) Interface {
+	return &implementation{ctl: fcfc.NewTestableController(informerFactory, flowcontrolClient, serverConcurrencyLimit, requestWaitLimit, queueSetFactory)}
+}
+
+func (impl *implementation) Run(stopCh <-chan struct{}) error {
+	return impl.ctl.Run(stopCh)
+}
+
+func (impl *implementation) Wait(ctx context.Context, requestDigest fcfc.RequestDigest) (bool, func()) {
+	startWaitingTime := time.Now()
+
+	// 1. classify the request
+	fsName, distinguisherMethod, plName, startFn := impl.ctl.Match(requestDigest)
+
+	// 2. early out for exempt
+	if startFn == nil {
+		klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s without delay", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+		startExecutionTime := time.Now()
+		return true, func() {
+			metrics.ObserveExecutionDuration(plName, fsName, time.Now().Sub(startExecutionTime))
+		}
+	}
+
+	// 3. computing hash
+	flowDistinguisher := computeFlowDistinguisher(requestDigest, distinguisherMethod)
+	hashValue := hashFlowID(fsName, flowDistinguisher)
+
+	// 4. queuing
+	execute, afterExecute := startFn(ctx, hashValue)
+
+	// 5. execute or reject
+	metrics.ObserveWaitingDuration(plName, fsName, strconv.FormatBool(execute), time.Now().Sub(startWaitingTime))
+	if !execute {
+		klog.V(7).Infof("Rejecting requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+		return false, func() {}
+	}
+	klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+	startExecutionTime := time.Now()
+	return execute, func() {
+		metrics.ObserveExecutionDuration(plName, fsName, time.Now().Sub(startExecutionTime))
+		afterExecute()
+	}
+}
+
+// computeFlowDistinguisher extracts the flow distinguisher according to the given method
+func computeFlowDistinguisher(rd fcfc.RequestDigest, method *fctypesv1a1.FlowDistinguisherMethod) string {
+	if method == nil {
+		return ""
+	}
+	switch method.Type {
+	case fctypesv1a1.FlowDistinguisherMethodByUserType:
+		return rd.User.GetName()
+	case fctypesv1a1.FlowDistinguisherMethodByNamespaceType:
+		return rd.RequestInfo.Namespace
+	default:
+		// this line shall never reach
+		panic("invalid flow-distinguisher method")
+	}
+}
+
+var hash32 bool
+
+// hashFlowID hashes the inputs into 64-bits
+func hashFlowID(fsName, fDistinguisher string) uint64 {
+	if hash32 {
+		return hash32FlowID(fsName, fDistinguisher)
+	}
+	return hash256FlowID(fsName, fDistinguisher)
+}
+
+func hash256FlowID(fsName, fDistinguisher string) uint64 {
+	hash := sha256.New()
+	hash.Write([]byte(fsName))
+	hash.Write([]byte{1})
+	hash.Write([]byte(fDistinguisher))
+	var sum [8]byte
+	hash.Sum(sum[:])
+	return uint64(sum[0])<<0 +
```

Comment:
Is this better than `import "encoding/binary"` and using e.g. `binary.LittleEndian.Uint64()`?

If we need to optimize this later we can consider the super fast unsafe cast, since byte order makes no difference at all here. For now let's stick to the readable thing.

Reply:
Temporary brain slip, I lost track of where to find that.

Reply:
fixed